File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

	static int n;
	static int m;
	static int[][] map;
	static boolean[][] visited;
	static int[] dx = {-1, 1, 0, 0};
	static int[] dy = {0, 0, -1, 1};

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		map = new int[n][m];
		visited = new boolean[n][m];

		for (int i = 0; i < n; i++) {
			String input = br.readLine();
			for (int j = 0; j < m; j++) {
				map[i][j] = input.charAt(j) - '0';
			}
		}

		bfs(0, 0);

		System.out.println(map[n - 1][m - 1]);
	}

	private static void bfs(int x, int y) {
		ArrayDeque<int[]> queue = new ArrayDeque<>();
		queue.offer(new int[]{x, y});
		visited[x][y] = true;

		while (!queue.isEmpty()) {
			int[] current = queue.poll();
			int curX = current[0];
			int curY = current[1];

			for (int i = 0; i < 4; i++) {
				int nx = curX + dx[i];
				int ny = curY + dy[i];

				if (0 <= nx && nx < n && 0 <= ny && ny < m && map[nx][ny] == 1 && !visited[nx][ny]) {
					queue.offer(new int[]{nx, ny});
					visited[nx][ny] = true;
					map[nx][ny] = map[curX][curY] + 1;
				}
			}
		}
	}
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

n, m = map(int, sys.stdin.readline().rstrip().split())
map = [list(map(int, list(sys.stdin.readline().rstrip()))) * m for _ in range(n)]
visited = [[False] * m for _ in range(n)]
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def bfs(x, y):
    queue = deque()
    queue.append([x, y])
    visited[x][y] = True

    while queue:
        current = queue.popleft()
        curX = current[0]
        curY = current[1]

        for i in range(4):
            nx = curX + dx[i]
            ny = curY + dy[i]

            if (
                0 <= nx
                and nx < n
                and 0 <= ny
                and ny < m
                and map[nx][ny] == 1
                and not visited[nx][ny]
            ):
                queue.append([nx, ny])
                visited[nx][ny] = True
                map[nx][ny] = map[curX][curY] + 1


bfs(0, 0)

print(map[n - 1][m - 1])
